File: sdk/python/tempmail_sdk/providers/tempmailc.py
"""
TempMailC 渠道 — https://tempmailc.com
"""

from typing import Any, Dict, List, Optional
from urllib.parse import quote

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "tempmailc"
API_BASE = "https://tempmailc.com/api/v1"
HEADERS = {"Accept": "application/json"}


def _flatten_list_message(raw: Dict[str, Any], recipient: str) -> Dict[str, Any]:
    return {
        "id": raw.get("id"),
        "from": raw.get("from"),
        "to": recipient,
        "subject": raw.get("subject", ""),
        "timestamp": raw.get("ts"),
        "read": raw.get("read"),
    }


def _fetch_message(email: str, message_id: str) -> Optional[Dict[str, Any]]:
    resp = tm_http.get(
        f"{API_BASE}/message?email={quote(email, safe='')}&msg_id={quote(message_id, safe='')}",
        headers=HEADERS,
        timeout=15,
    )
    if not resp.ok:
        return None
    data = resp.json()
    if not isinstance(data, dict) or data.get("ok") is False:
        return None
    return data


def generate_email() -> EmailInfo:
    resp = tm_http.get(f"{API_BASE}/new", headers=HEADERS, timeout=15)
    resp.raise_for_status()
    data = resp.json()
    email = str(data.get("email") or "").strip() if isinstance(data, dict) else ""
    if not isinstance(data, dict) or not data.get("ok") or not email or "@" not in email:
        raise ValueError("tempmailc: invalid new email response")
    return EmailInfo(channel=CHANNEL, email=email)


def get_emails(email: str) -> List[Email]:
    address = (email or "").strip()
    if not address:
        raise ValueError("tempmailc: empty email")

    resp = tm_http.get(
        f"{API_BASE}/inbox?email={quote(address, safe='')}",
        headers=HEADERS,
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, dict) or not data.get("ok"):
        raise ValueError("tempmailc: inbox response failed")

    rows = data.get("messages")
    if not isinstance(rows, list):
        return []

    emails: List[Email] = []
    for item in rows:
        if not isinstance(item, dict):
            continue
        message_id = str(item.get("id") or "").strip()
        detail = _fetch_message(address, message_id) if message_id else None
        emails.append(normalize_email(detail or _flatten_list_message(item, address), address))
    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/tempmailc.rb
# frozen_string_literal: true

module TempmailSdk
  module Providers
    # TempMailC 渠道实现
    # API: https://tempmailc.com/api/v1
    module Tempmailc
      CHANNEL = "tempmailc"
      API_BASE = "https://tempmailc.com/api/v1"
      HEADERS = { "Accept" => "application/json" }.freeze

      module_function

      # 扁平化列表消息（无正文时使用）
      def flatten_list_message(raw, recipient)
        {
          "id" => raw["id"],
          "from" => raw["from"],
          "to" => recipient,
          "subject" => raw["subject"] || "",
          "timestamp" => raw["ts"],
          "read" => raw["read"]
        }
      end

      # 获取单封邮件详情
      def fetch_message(email, message_id)
        resp = Http.get(
          "#{API_BASE}/message?email=#{URI.encode_www_form_component(email)}&msg_id=#{URI.encode_www_form_component(message_id)}",
          headers: HEADERS, timeout: 15
        )
        return nil unless resp.ok?

        data = resp.json
        return nil unless data.is_a?(Hash) && data["ok"] != false

        data
      rescue StandardError
        nil
      end

      # 创建临时邮箱
      # @return [EmailInfo]
      def generate_email
        resp = Http.get("#{API_BASE}/new", headers: HEADERS, timeout: 15)
        resp.raise_for_status
        data = resp.json
        email = data.is_a?(Hash) ? data["email"].to_s.strip : ""
        raise "tempmailc: 无效的响应" unless data.is_a?(Hash) && data["ok"] && !email.empty? && email.include?("@")

        EmailInfo.new(channel: CHANNEL, email: email)
      end

      # 获取邮件列表
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(email)
        address = email.to_s.strip
        raise "tempmailc: 邮箱地址为空" if address.empty?

        resp = Http.get(
          "#{API_BASE}/inbox?email=#{URI.encode_www_form_component(address)}",
          headers: HEADERS, timeout: 15
        )
        resp.raise_for_status
        data = resp.json
        raise "tempmailc: inbox 响应失败" unless data.is_a?(Hash) && data["ok"]

        rows = data["messages"]
        return [] unless rows.is_a?(Array)

        rows.select { |item| item.is_a?(Hash) }.map do |item|
          mid = item["id"].to_s.strip
          detail = mid.empty? ? nil : fetch_message(address, mid)
          Normalize.normalize_email(detail || flatten_list_message(item, address), address)
        end
      end
    end
  end
end
